Sources/OTel/Metrics/MetricStore/OTelMetricRegistry.swift

```diff
@@ -0,0 +1,284 @@
+            storage.gauges[identifier] = [attributes: newInstrument]
+            return newInstrument
+        }
+    }
+
+    func makeDurationHistogram(name: String, unit: String? = nil, description: String? = nil, attributes: Set<Attribute> = [], buckets: [Duration]) -> DurationHistogram {
+        storage.withLockedValue { storage in
+            let identifier = InstrumentIdentifier.histogram(name: name, unit: unit, description: description)
+            if var existingInstruments = storage.durationHistograms[identifier] {
+                if let existingInstrument = existingInstruments[attributes] {
+                    return existingInstrument
+                }
+                let newInstrument = DurationHistogram(name: name, unit: unit, description: description, attributes: attributes, buckets: buckets)
+                existingInstruments[attributes] = newInstrument
+                return newInstrument
+            }
+            storage.register(identifier, forName: name)
+            let newInstrument = DurationHistogram(name: name, unit: unit, description: description, attributes: attributes, buckets: buckets)
+            storage.durationHistograms[identifier] = [attributes: newInstrument]
+            return newInstrument
+        }
+    }
+
+    func makeValueHistogram(name: String, unit: String? = nil, description: String? = nil, attributes: Set<Attribute> = [], buckets: [Double]) -> ValueHistogram {
+        storage.withLockedValue { storage in
+            let identifier = InstrumentIdentifier.histogram(name: name, unit: unit, description: description)
+            if var existingInstruments = storage.valueHistograms[identifier] {
+                if let existingInstrument = existingInstruments[attributes] {
+                    return existingInstrument
+                }
+                let newInstrument = ValueHistogram(name: name, unit: unit, description: description, attributes: attributes, buckets: buckets)
+                existingInstruments[attributes] = newInstrument
+                return newInstrument
+            }
+            storage.register(identifier, forName: name)
+            let newInstrument = ValueHistogram(name: name, unit: unit, description: description, attributes: attributes, buckets: buckets)
+            storage.valueHistograms[identifier] = [attributes: newInstrument]
+            return newInstrument
+        }
+    }
+
+    func unregisterCounter(_ counter: Counter) {
+        let identifier = counter.instrumentIdentifier
+        self.storage.withLockedValue { storage in
+            if var existingInstrument = storage.counters[identifier] {
+                existingInstrument.removeValue(forKey: counter.attributes)
+                if existingInstrument.isEmpty {
+                    storage.counters.removeValue(forKey: identifier)
+                    storage.unregister(identifier, forName: identifier.name)
+                }
+            }
```

Comment:
This pattern of removing and clearing the storage is used in all `unregister` methods, as far as I can tell. Do you think it's worth refactoring into a method?

Something like:

```swift
private func unregister<Instrument: IdentifiableInstrument>(_ keyPath: KeyPath<Storage, [InstrumentIdentifier: Instrument]>) {
    let identifier = instrument.identifier
    storage.withLockedValue { storage in
        if var existingInstrument = storage[keyPath: keyPath][identifier] {
            existingInstrument.removeValue(forKey: instrument.attributes)
            if existingInstrument.isEmpty {
                storage[keyPath: keyPath].removeValue(forKey: identifier)
                storage.unregister(identifier, forName: identifier.name)
            }
        }
    }
}

func unregisterCounter(_ counter: Counter) {
    self.unregister(\.counters, counter)
}
```

This would require adding a new protocol or extending `IdentifiableInstrument` to also require `attributes`. Please note that I haven't compiled this code snippet, it's just a brain dump.

Reply:
IMHO I don't think this is worth it. I _did_ factor out the duplicate registration logic as it is independent of the instrument kind, but that needed to be pushed into the storage type too because we're holding the lock.

I'm happy to revisit this as an implementation detail if we're happy with the API, so we don't block this PR, which is holding up the others. WDYT?

Reply:
Yeah, fair point. Let's merge first and optionally look back at it in the future 👍